ledger/streaming/client/src/backfill.rs

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! A BlockStream that backfills missing indices from a given BlockFetcher.
+
+use displaydoc::Display;
+use futures::{FutureExt, Stream, StreamExt};
+use mc_common::logger::{log, Logger};
+use mc_ledger_streaming_api::{BlockFetcher, BlockStream, BlockStreamComponents, Result};
+use mc_transaction_core::BlockIndex;
+use std::{pin::Pin, sync::Arc};
+
+/// A [BlockStream] that backfills another [BlockStream] using a [BlockFetcher].
+#[derive(Debug, Display)]
+pub struct BackfillingStream<S: BlockStream + 'static, F: BlockFetcher + 'static> {
+    upstream: S,
+    fetcher: Arc<F>,
+    logger: Logger,
+}
+
+impl<S: BlockStream + 'static, F: BlockFetcher + 'static> BackfillingStream<S, F> {
+    /// Instantiate a [BackfillingStream].
+    pub fn new(upstream: S, fetcher: F, logger: Logger) -> Self {
+        Self {
+            upstream,
+            fetcher: Arc::new(fetcher),
+            logger,
+        }
+    }
+}
+
+impl<S: BlockStream + 'static, F: BlockFetcher + 'static> BlockStream for BackfillingStream<S, F> {
+    type Stream = impl Stream<Item = Result<BlockStreamComponents>>;
+
+    fn get_block_stream(&self, starting_height: u64) -> Result<Self::Stream> {
+        self.upstream
+            .get_block_stream(starting_height)
+            .map(|upstream| {
+                backfill_stream(
+                    upstream,
+                    starting_height,
+                    self.fetcher.clone(),
```

Comment:
This is a clone, but it's an Arc clone where the actual arg is attached to the BackfillingStream object. This could potentially cause lifetime issues. Likely better to make the arc within the stream.

```suggestion
                    Arc::new(self.fetcher.clone()),
```

Reply:
This is intentional -- if the `BackfillingStream` somehow goes out of scope while one of its returned streams is running, those streams will retain a copy of the fetcher.

What you're suggesting would clone the fetcher, which is not what I want here.